integrations/jta/jdbc/src/main/java/io/helidon/integrations/jta/jdbc/LocalXAResource.java

```diff
@@ -0,0 +1,1031 @@
+     */
+
+
+    private static final Logger LOGGER = Logger.getLogger(LocalXAResource.class.getName());
+
+    private static final Xid[] EMPTY_XID_ARRAY = new Xid[0];
+
+    // package-protected for testing only.
+    static final ConcurrentMap<Xid, Association> ASSOCIATIONS = new ConcurrentHashMap<>();
+
+
+    /*
+     * Instance fields.
+     */
+
+
+    private final Function<? super Xid, ? extends Connection> connectionFunction;
+
+    private final ExceptionConverter exceptionConverter;
+
+
+    /*
+     * Constructors.
+     */
+
+
+    /**
+     * Creates a new {@link LocalXAResource}.
+     *
+     * @param connectionFunction a {@link Function} that accepts a {@link Xid} (supplied by the {@link #start(Xid, int)}
+     * method) and returns a {@link Connection} to associate with the global transaction; must not be {@code null}; must
+     * never return {@code null}; must be safe for concurrent use by multiple threads; will never be invoked with a
+     * {@code null} {@link Xid}
+     *
+     * @param exceptionConverter a {@link ExceptionConverter} that accepts a {@link XARoutine} and a {@link
+     * SQLException} and converts the {@link SQLException} to an <em>appropriate</em> {@link XAException} following the
+     * rules defined by the <a href="https://pubs.opengroup.org/onlinepubs/009680699/toc.pdf">XA Specification</a> as
+     * interpreted by the specification of the {@code javax.transaction.xa} package and its classes; may be {@code null}
+     * in which case a default implementation will be used instead
+     *
+     * @see #start(Xid, int)
+     */
+    LocalXAResource(Function<? super Xid, ? extends Connection> connectionFunction, ExceptionConverter exceptionConverter) {
+        super();
+        this.connectionFunction = Objects.requireNonNull(connectionFunction, "connectionFunction");
+        this.exceptionConverter = exceptionConverter == null ? LocalXAResource::convert0 : exceptionConverter;
+    }
+
+
+    /*
+     * Instance methods.
+     */
+
+
+    @Override // XAResource
+    public void start(Xid xid, int flags) throws XAException {
+        if (LOGGER.isLoggable(Level.FINER)) {
+            LOGGER.entering(this.getClass().getName(), "start", new Object[] {xid, flagsToString(flags)});
+        }
+        requireNonNullXid(xid);
+
+        BiFunction<Xid, Association, Association> remappingFunction;
+        switch (flags) {
+        case TMJOIN:
+            remappingFunction = LocalXAResource::join;
+            break;
+        case TMNOFLAGS:
+            remappingFunction = this::start;
+            break;
+        case TMRESUME:
+            remappingFunction = LocalXAResource::resume;
+            break;
+        default:
+            // Bad flags.
+            throw (XAException) new XAException(XAER_INVAL)
+                .initCause(new IllegalArgumentException("xid: " + xid + "; flags: " + flagsToString(flags)));
+        }
+
+        this.computeAssociation(XARoutine.START, xid, remappingFunction);
```

Comment:
We select the appropriate `BiFunction` to pass in to the `computeAssociation` method.  As hopefully you can see above, it's determined by what switches the transaction manager has passed when it called this `start` method.

Reply:
"Remapping" in the sense documented here: https://docs.oracle.com/en/java/javase/19/docs/api/java.base/java/util/concurrent/ConcurrentMap.html#compute(K,java.util.function.BiFunction)

I use [`ConcurrentMap#compute(Object, BiFunction)`](https://docs.oracle.com/en/java/javase/19/docs/api/java.base/java/util/concurrent/ConcurrentMap.html#compute(K,java.util.function.BiFunction)) because for some invocations of, say, `start` the `Association` (the value) will not exist, and will need to be created, and in other invocations the `Association` will exist (i.e. in the case of joining).

Reply:
Ah, OK. I was trying to make "remapping" fit the vocabulary of XA rather than of the data structure API you're using.